### PR TITLE
fix: physically delete retained release histories

### LINKED ISSUE #5640: Release history retention only soft-deletes records and does not reduce table/index pressure
### Describe the bug

PR #4813 added the release history retention feature via `apollo.release-history.retention.size` and `apollo.release-history.retention.size.override`. The stated purpose is to avoid excessive database pressure caused by too many release history records.

However, the current cleanup path only performs **soft delete** for `ReleaseHistory` and the associated `Release` records. As a result, the retention feature hides old records from application queries, but it does not actually reduce table size or secondary-index size, and the old rows remain in MySQL.

Related PR / discussions:

- #4813
- #5260
- #5531

### Current behavior

`ReleaseHistoryService.cleanReleaseHistory(...)` collects old release history records in batches of 100 and then calls:

```java
releaseHistoryRepository.deleteAll(cleanReleaseHistoryList);
releaseRepository.deleteAllById(releaseIds);
```

But both entities are configured with Hibernate `@SQLDelete`:

```java
@SQLDelete(
    sql = "Update `ReleaseHistory` set IsDeleted = true, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000) where Id = ?")
@SQLRestriction("`IsDeleted` = false")
public class ReleaseHistory extends BaseEntity {
```

```java
@SQLDelete(
    sql = "Update `Release` set IsDeleted = true, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000) where Id = ?")
@SQLRestriction("`IsDeleted` = false")
public class Release extends BaseEntity {
```

So the actual cleanup SQL is effectively:

```sql
UPDATE `ReleaseHistory`
SET IsDeleted = true,
    DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000)
WHERE Id = ?;
```

and:

```sql
UPDATE `Release`
SET IsDeleted = true,
    DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000)
WHERE Id = ?;
```

### Why this is a problem

For operators, the expected value of release history retention is to reduce the amount of old release data that keeps accumulating in MySQL. With the current soft-delete implementation:

- physical rows are still retained;
- secondary-index entries are still retained;
- table size, backup size, and index size are not reduced;
- MySQL buffer pool / storage pressure is not materially improved;
- already soft-deleted records are filtered by `@SQLRestriction("IsDeleted = false")`, so future retention cleanup will not pick them up again for physical deletion;
- users still need DBA/manual purge jobs such as deleting `IsDeleted = 1` rows in batches.

This means the feature works for UI/application visibility, but it does not fully satisfy the retention goal of reducing database pressure.

### Expected behavior

Release history retention should either:

1. physically delete stale `ReleaseHistory` and associated `Release` rows in the retention cleanup path, or
2. provide a separate built-in purge job for rows already marked `IsDeleted = true`, or
3. clearly document that retention only soft-deletes rows and that operators must run their own physical purge job to reclaim table/index space.

### Suggested fix

Keep `@SQLDelete` for normal business delete semantics, but bypass it only for the retention cleanup path by adding explicit physical delete repository methods, for example:

```java
@Modifying(clearAutomatically = true, flushAutomatically = true)
@Query(value = "DELETE FROM `ReleaseHistory` WHERE `Id` IN (:ids)", nativeQuery = true)
int deletePhysicallyByIdIn(@Param("ids") Collection<Long> ids);
```

and:

```java
@Modifying(clearAutomatically = true, flushAutomatically = true)
@Query(value = "DELETE FROM `Release` WHERE `Id` IN (:ids)", nativeQuery = true)
int deletePhysicallyByIdIn(@Param("ids") Collection<Long> ids);
```

Then `ReleaseHistoryService.cleanReleaseHistory(...)` can use these methods for retention cleanup only.

A test should verify physical row count with native SQL, because repository `count()` is affected by `@SQLRestriction` and would not distinguish soft delete from physical delete.

### Version

Observed in the current codebase and the release-history retention implementation introduced by #4813.

### PR DESCRIPTION
**Pull request by klboke**
## What
Fix release history retention cleanup so stale `ReleaseHistory` rows and their associated `Release` rows are physically deleted instead of only being soft-deleted through Hibernate `@SQLDelete`.

## Why
`apollo.release-history.retention.size` is intended to reduce accumulated release-history database pressure. The existing cleanup path called repository delete methods that honor entity soft-delete mappings, so old rows remained in MySQL tables and indexes.

## How
- Add retention-only native delete repository methods for `ReleaseHistory` and `Release`.
- Use those physical delete methods inside the existing retention cleanup transaction.
- Extend `ReleaseHistoryServiceTest` with native SQL row-count assertions so the test distinguishes physical deletion from soft deletion.

## Which issue(s) this PR fixes:
Fixes #5640

## Brief changelog
Fix release history retention cleanup to physically delete stale release history and release rows.

Follow this checklist to help us incorporate your contribution quickly and easily:

- [ ] Read the [Contributing Guide](https://github.com/apolloconfig/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [ ] Run `mvn clean test` to make sure this pull request does not break anything.
- [x] Run `mvn spotless:apply` to format your code.
- [x] Update the [`CHANGES` log](https://github.com/apolloconfig/apollo/blob/master/CHANGES.md).

## Tests
- `./mvnw -pl apollo-biz -am -Dtest=ReleaseHistoryServiceTest -Dsurefire.failIfNoSpecifiedTests=false test`
- `./mvnw -pl apollo-biz spotless:check`

## Summary by CodeRabbit

* **Bug Fixes**
  * Updated release-history retention cleanup to physically delete eligible retained history and physically remove only releases that are no longer referenced.
  * Improved cleanup batching/termination and ensured physical row-count stability during transactional rollback.
* **Documentation**
  * Added release note entry and a manual SQL migration runbook (prechecks, restore, and bounded purge flow).
* **Tests**
  * Enhanced retention cleanup and migration-script tests with physical row-count assertions, idempotency checks, and retention boundary scenarios.
* **Performance**
  * Added/updated indexes on gray release rules to accelerate retention/cleanup lookups.

### CODE REVIEW

**Review by nobodyiam — CHANGES_REQUESTED** [review]
The retention cleanup now physically deletes rows on the happy path, but the current release deletion is not reference-safe and can irreversibly delete releases that are still in use.

Please address the following blocking items:

1. In `ReleaseHistoryService.cleanReleaseHistory(...)`, do not physically delete a `Release` merely because one pruned history references it. `ReleaseId` is reused by later histories during rollback and gray-rule updates, so the release may still be referenced by retained `ReleaseHistory` rows or an active `GrayReleaseRule`. Please make the delete reference-safe and add regression coverage for a retained history reusing an older release ID.
2. Since this PR declares `Fixes #5640`, please also provide a safe purge/migration path for rows already soft-deleted by previous retention runs, or clearly document and scope the remaining migration requirement. The current JPA finder is still filtered by `@SQLRestriction`, so those existing rows are never selected.

The existing happy-path and transaction-rollback tests pass locally, and the current CI checks are green, but the shared-release regression above currently fails.

File: apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ReleaseHistoryService.java [hunk 1]
```diff
@@ -201,14 +201,19 @@ private void cleanReleaseHistory(ReleaseHistory cleanRelease) {
       List<ReleaseHistory> cleanReleaseHistoryList = releaseHistoryRepository
           .findFirst100ByAppIdAndClusterNameAndNamespaceNameAndBranchNameAndIdLessThanEqualOrderByIdAsc(
               appId, clusterName, namespaceName, branchName, maxId.get());
+      if (cleanReleaseHistoryList.isEmpty()) {
+        break;
+      }
```
[the remark this answers is not in the document]
Comment (klboke) [reply apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ReleaseHistoryService.java:207]: Addressed in 496dbeb0. Existing soft-deleted rows are handled as a one-time manual migration instead of adding a legacy-row scan to every retention cleanup. The new `scripts/sql/migration/release-history-retention` directory contains two prechecks plus restore and physical-purge scripts, with ordering, backup, batching, and idempotency guidance. `ReleaseHistoryRetentionMigrationSqlTest` directly executes those four repository scripts and covers discovery, repair, purge, conflicts, reference protection, and repeated execution. The scripts were also validated against MySQL 5.7.

File: apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ReleaseHistoryService.java [hunk 2]
```diff
@@ -201,14 +201,19 @@ private void cleanReleaseHistory(ReleaseHistory cleanRelease) {
...
+      List<Long> releaseHistoryIds =
+          cleanReleaseHistoryList.stream().map(ReleaseHistory::getId).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (klboke) [reply apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ReleaseHistoryService.java:209]: Addressed in 496dbeb0. Legacy `IsDeleted = true` histories are intentionally repaired and purged through the new one-time manual migration under `scripts/sql/migration/release-history-retention`, rather than being scanned on every runtime cleanup. It provides two prechecks, a referenced-release restore step, and a batched physical purge step. Automated tests load and execute the actual four SQL files, including idempotency and conflict/reference protection, and the scripts were validated on MySQL 5.7.

File: apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ReleaseHistoryService.java [hunk 3]
```diff
@@ -201,14 +201,19 @@ private void cleanReleaseHistory(ReleaseHistory cleanRelease) {
...
       Set<Long> releaseIds = cleanReleaseHistoryList.stream().map(ReleaseHistory::getReleaseId)
           .collect(Collectors.toSet());
 
       transactionManager.execute(new TransactionCallbackWithoutResult() {
         @Override
         protected void doInTransactionWithoutResult(TransactionStatus status) {
-          releaseHistoryRepository.deleteAll(cleanReleaseHistoryList);
-          releaseRepository.deleteAllById(releaseIds);
+          releaseHistoryRepository.deletePhysicallyByIdIn(releaseHistoryIds);
+          releaseRepository.deletePhysicallyByIdIn(releaseIds);
```
[the remark this answers is not in the document]
Comment (klboke) [reply apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ReleaseHistoryService.java]: Addressed in 496dbeb0. Cleanup now collects both `releaseId` and `previousReleaseId`, deletes the selected history batch first, and then physically deletes only releases that are no longer referenced by any remaining `ReleaseHistory` row or active `GrayReleaseRule`. The two history checks are separate `NOT EXISTS` clauses so the existing `IX_ReleaseId` and `IX_PreviousReleaseId` indexes can be used. Regression tests cover a shared release ID, a retained previous-release reference, and an active gray-release-rule reference.

**Review by nobodyiam — CHANGES_REQUESTED** [review]
The runtime reference-safety fix looks good, and the shared `ReleaseId` / `PreviousReleaseId` regression coverage now passes. However, the new legacy migration still has two blocking data-boundary issues:

1. `04-purge-soft-deleted-rows.sql` deletes every `ReleaseHistory` row with `IsDeleted = true` and then every unreferenced soft-deleted `Release`. The same soft-delete shape is also produced by normal namespace deletion through `NamespaceService.deleteNamespace(...)`, so the script cannot currently distinguish retention leftovers from normal business soft deletes. Please scope the purge to data that can be safely identified as retention-generated, or explicitly redesign and document it as a broader destructive migration requiring separate maintainer approval.
2. The runtime guard and migration scripts treat every `GrayReleaseRule` with `IsDeleted = false` as active. Deleted and merged branch tombstones remain non-deleted but have `BranchStatus = DELETED/MERGED`; the restore script currently reactivates a soft-deleted release referenced only by such a tombstone. Please require `BranchStatus = ACTIVE` consistently in the runtime query, prechecks, restore, and purge paths, and add regression tests for inactive branch rules.

The current 10 positive tests and all required CI checks pass, but focused negative tests for these two scenarios fail on head `3f18b9f52`.

**Review by nobodyiam — CHANGES_REQUESTED** [review]
Thanks for the update. The previous shared-release reference issue and the inactive gray-rule issue are now addressed, and the added index/test coverage looks good.

One blocking data-safety edge case remains in the namespace-incarnation scoping.

The migration scripts classify rows as belonging to the current namespace incarnation with:

    DataChange_LastTime >= Namespace.DataChange_CreatedTime

Both columns have second-level timestamp precision. If a namespace is deleted and recreated with the same app/cluster/namespace tuple within the same second, rows from the previous incarnation can have a `DataChange_LastTime` equal to the new Namespace row `DataChange_CreatedTime`. Those normal namespace-deletion rows then pass the `>=` predicate and are physically purged, despite the README stating that previous namespace incarnations are excluded.

I reproduced this with a focused regression that executes the current `04-purge-soft-deleted-rows.sql`: when the old soft-deleted release/history rows and the recreated namespace use the same-second boundary, the old history is deleted.

Please use an unambiguous incarnation boundary—e.g. the millisecond-resolution `DeletedAt`, or another approach that cannot collide at second precision—apply it consistently across the precheck/restore/purge scripts, and add a same-second delete/recreate regression test.

Non-blocking: the newly added migration test class should preferably use JUnit 5, which is the repository default for new tests.

**Review by nobodyiam — APPROVED** [review]
Thanks for the update. I re-reviewed the current head and the previous same-second namespace-recreation blocker is resolved.

The migration now consistently uses the millisecond-resolution `DeletedAt` boundary to distinguish the current active namespace incarnation from prior deleted incarnations. The new regression covers same-second deletion/recreation and verifies that previous-incarnation rows remain untouched while current-incarnation retention rows are purged. The migration test was also converted to JUnit 5, and the ReleaseHistory reference lookups now use the existing indexes explicitly.

I locally verified the focused migration, service, and SQL-converter tests (17 tests total), and also executed all four migration scripts against a temporary MySQL instance with the same-second fixture. The expected rows were preserved and purged correctly.

The documented `ActiveReleaseKeyConflicts = 0` precondition, stopped-service requirement, transactional failure behavior, and regression coverage provide a sufficient safety gate for release-key conflicts.

No blocking findings remain from my side.
